### PR TITLE
Tutorial: Add tutorial on just-in-time Holoscan debugging

### PR DESCRIPTION
**Pull request by tbirdso**
Adds a tutorial to walk through setting up just-in-time debugging with GDB or `pdb` debugging tools for a Holoscan SDK application using HoloHub infrastructure.

Includes scripts for runnable examples illustrating just-in-time debugging. Streamlined instructions for v2.3 rely on debugging symbols to be distributed in the Holoscan SDK NGC development container, while general instructions for <=v2.2 are also included.

Depends on planned changes to the Holoscan SDK v2.3 along with fixes for the Endoscopy Tool Tracking app in https://github.com/nvidia-holoscan/holohub/pull/412. I've focused the tutorial on headless tools (GDB, pdb) for now, would be interested to add a section on VS Code IDE debugging tools following merge of https://github.com/nvidia-holoscan/holohub/pull/425 .

### CODE REVIEW

**Comment by tbirdso** [comment]
Ready for review, but should not be merged until the Holoscan SDK v2.3 container is available.

File: tutorials/holoscan_debugging/README.md [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+```
+
+#### How can I attach to a running Holoscan SDK session?
+
+Do the following to attach to a HoloHub application:
+
+1. Launch the container with root permissions and start the process in the background:
+```bash
+./dev_container launch --img holohub:debugging --as_root
+>>> ./run launch endoscopy_tool_tracking &     # then Ctrl+C to return to your interactive shell
```
Comment (mocsharp) [inline tutorials/holoscan_debugging/README.md:130]: Any reason to include `>>>` here? Does it mean inside the container?

Comment (mocsharp) [inline tutorials/holoscan_debugging/README.md:130]: Maybe we can make the comment an explicit step so the users won't miss it?

```suggestion
>>> ./run launch endoscopy_tool_tracking &    

2. A Holoviz window should open to play a video of an endoscopy surgery.  Press Ctrl+C on the keyboard to return to your interactive shell.
```

File: tutorials/holoscan_debugging/README.md [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+
+2. Find the process ID (PID) of the running application:
+```bash
+>>> ps -ef | grep endoscopy_tool_tracking
+user+     292     203 28 13:17 pts/9    00:00:04 /workspace/holohub/build/endoscopy_tool_tracking/applications/endoscopy_tool_tracking/cpp/endoscopy_tool_tracking --data /workspace/holohub/data/endoscopy
+```
+
+3. Attach to the running process with `gdb -p`:
+```bash
+>>> gdb -p 292
```
Comment (mocsharp) [inline tutorials/holoscan_debugging/README.md:141]: I got the following error, and I assume that's because we don't include GXF extension debug symbols. Do we want to add a note to this?

```bash
warning: Could not load shared library symbols for gxf_extensions/lstm_tensor_rt_inference/libgxf_lstm_tensor_rt_inference.so.
```

File: tutorials/holoscan_debugging/README.md [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+This command launches the Python version of the Endoscopy Tool Tracking application with a breakpoint set on the
+very first line. From this point you can set additional breakpoints, inspect application state, and control program
+execution.
+
+Although launched from Python, calls to `import holoscan` will load Python-wrapped C++ Holoscan SDK libraries
+along with their debugging symbols. You can set breakpoints in the corresponding Holoscan SDK C++ source code after each
+breakpoint launches. For instance:
+```py
+# Load the Holoscan core module in Python
+-> from holoscan.core import Application
```
Comment (mocsharp) [inline tutorials/holoscan_debugging/README.md:203]: I'm a bit confused on this line, should the user enter this line into the (Pdb) prompt? If so, should we change it to:

```suggestion
(Pdb) from holoscan.core import Application
```
I did notice that if I hit `n` twice, I will see the following, so I'm not sure what to do here.
```
> /workspace/holohub/applications/endoscopy_tool_tracking/python/endoscopy_tool_tracking.py(19)<module>()
-> from holoscan.core import Application
```

File: tutorials/holoscan_debugging/README.md [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+(Pdb) n
+> /workspace/holohub/applications/endoscopy_tool_tracking/python/endoscopy_tool_tracking.py(20)<module>()
+# Core debugging symbols are now available we can set breakpoints in C++ code.
+(Pdb) break /workspace/holoscan-sdk/src/core/application.cpp:172
+Breakpoint 1 at /workspace/holoscan-sdk/src/core/application.cpp:172
```
Comment (mocsharp) [inline tutorials/holoscan_debugging/README.md:208]: Btw, if I continue to run both the C++ and the Python debugging session, I see the following:

```
[error] [graph_entity.hpp:646] Typename [nvidia::holoscan::lstm_tensor_rt_inference::TensorRtInference] not found. Is this type registered?
[error] [graph_entity.cpp:159] Graph Entity failed to create codelet of type [nvidia::holoscan::lstm_tensor_rt_inference::TensorRtInference]
[error] [gxf_executor.cpp:1399] Exception occurred during initialization of operator: 'lstm_inferer' - Failed to add codelet of type nvidia::holoscan::lstm_tensor_rt_inference::TensorRtInference
```

File: tutorials/holoscan_debugging/README.md [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+If you are using a legacy Holoscan SDK container (earlier than v2.3) for your development, your container does not come with
+debugging symbols pre-packaged. However, you can rebuild Holoscan SDK from source to enable interactive debugging.
+We provide utilities as part of Holoscan SDK open source code to help you generate debugging
+symbols that will allow you to use tools such as `gdb` and `pdb` with legacy Holoscan SDK.
+
+The following script provides the necessary steps to rebuild Holoscan SDK version with debugging symbols and then
+set up for debugging with `gdb`:
+
+```bash
+./tutorials/holoscan_debugging/debug_legacy.sh <build_type> <sdk_tag> <ngc_tag>
```
Comment (mocsharp) [inline tutorials/holoscan_debugging/README.md:233]: Can we provide an example with actual values? :pray:  
Maybe for convenience, links to the tags and NGC page can also be added so the user can easily locate available tags.

https://github.com/nvidia-holoscan/holoscan-sdk/tags
https://catalog.ngc.nvidia.com/orgs/nvidia/teams/clara-holoscan/containers/holoscan/tags

File: tutorials/holoscan_debugging/README.md [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+
+## Debugging Symbols for Legacy Holoscan SDK Versions
+
+In Holoscan SDK v2.3 we began packaging debugging symbols as part of the libraries distributed in
+[NGC Holoscan Containers](https://catalog.ngc.nvidia.com/orgs/nvidia/teams/clara-holoscan/containers/holoscan), with the goal
+of improving ease of development and debugging. This change comes at a small cost to size and performance of the Holoscan SDK
+binary distribution. But what about older versions of Holoscan SDK containers?
+
+If you are using a legacy Holoscan SDK container (earlier than v2.3) for your development, your container does not come with
+debugging symbols pre-packaged. However, you can rebuild Holoscan SDK from source to enable interactive debugging.
```
Comment (mocsharp) [inline tutorials/holoscan_debugging/README.md:225]:
```suggestion
debugging symbols pre-packaged. However, you can rebuild the Holoscan SDK from the source to enable interactive debugging.
```

**Review by mocsharp — APPROVED** [review]
Thanks for this detailed tutorial.

File: tutorials/holoscan_debugging/README.md (hunk 2 above)
Comment (tbirdso) [reply tutorials/holoscan_debugging/README.md:155]: Fixed, this was an issue with the working directory. Thanks for reporting

File: tutorials/holoscan_debugging/README.md (hunk 3 above)
Comment (tbirdso) [reply tutorials/holoscan_debugging/README.md]: The line in question was output from Pdb to the console, but I see how it was confusing. I've reworked the example for clarity (hopefully).

File: tutorials/holoscan_debugging/README.md (hunk 4 above)
Comment (tbirdso) [reply tutorials/holoscan_debugging/README.md]: Fixed (working directory issue) -- thanks!

File: tutorials/holoscan_debugging/README.md (hunk 5 above)
Comment (tbirdso) [reply tutorials/holoscan_debugging/README.md:249]: The script uses default values for v2.2.0. Would it be more helpful to remove the generic arguments here or to replace them with the default values?

Comment (mocsharp) [reply tutorials/holoscan_debugging/README.md:249]: I think what you have looks good, but providing an example would make it easier to follow because not everyone knows what `ngc_tag` means. A few links to let the users know where to find the values would help, too.

File: tutorials/holoscan_debugging/README.md [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+to get started.
+
+### Prerequisites
+
+The steps for getting started with `gdb` depend on how you are consuming Holoscan SDK.
+- We encourage using Holoscan SDK containers from NGC for development and we take this approach for most of the tutorial. If you are using an NGC container for __Holoscan SDK v2.3.0 or later__, you already have access
+  to debug symbols and can get started right away.
+- If you are using an older Holoscan SDK container from NGC, or if you are consuming Holoscan SDK through another means such as Debian packages, Python wheels, or custom installation, you will need to build
+Holoscan SDK with debugging symbols in order to step through Holoscan SDK code during debugging. Jump to the [Legacy Holoscan SDK](#debugging-symbols-for-legacy-holoscan-sdk-versions) section to get started.
+
```
Comment (zhijinl) [inline tutorials/holoscan_debugging/README.md:80]: Aren't these prerequisites true independent of which tool to use (gdb, pdb, legacy)?
I'd suggest moving these to end of section "What are debugging symbols?", as general per-requisites.

File: tutorials/holoscan_debugging/README.md [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+When the `python` argument is provided, the script builds the Endoscopy Tool Tracking application with Python bindings
+and initiates debugging with GDB from the Python script entrypoint. Once symbols are loaded in GDB, we can set breakpoints
+and inspect the underlying state of Holoscan SDK C++ operator implementations at runtime.
+
+From this point we recommend referring to the [GDB Manual](https://sourceware.org/gdb/current/onlinedocs/gdb.html/) or
+online tutorials to get started with interactive debugging commands.
+
+### Frequently Asked Questions and Troubleshooting
+
+#### How can I verify that Holoscan SDK debugging symbols have loaded in `gdb`?
```
Comment (zhijinl) [inline tutorials/holoscan_debugging/README.md:124]:
```suggestion
#### How can I verify that Holoscan SDK debugging symbols have been loaded in `gdb`?
```

File: tutorials/holoscan_debugging/README.md [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+```
+
+You can also add a breakpoint directly in the `.py` source code before running a program by adding a `breakpoint()` statement.
+
+From here we recommend referring to [`pdb` documentation](https://docs.python.org/3/library/pdb.html) for common
+commands and debugging strategies.
+
+## Debugging Symbols for Legacy Holoscan SDK Versions
+
+In Holoscan SDK v2.3 we began packaging debugging symbols as part of the libraries distributed in
```
Comment (zhijinl) [inline tutorials/holoscan_debugging/README.md:235]:
```suggestion
Starting from Holoscan SDK v2.3, we package debugging symbols as part of the libraries distributed in
```

File: tutorials/holoscan_debugging/README.md [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+
+The following script provides the necessary steps to rebuild Holoscan SDK version with debugging symbols and then
+set up for debugging with `gdb`:
+
+```bash
+./tutorials/holoscan_debugging/debug_legacy.sh <build_type> <sdk_tag> <ngc_tag>
+```
+
+The script does the following:
+1. Builds the specified Holoscan SDK version with the specified build type in a temporary tutorial folder.
```
Comment (zhijinl) [inline tutorials/holoscan_debugging/README.md:253]: For build type: do we prefer `rel-debug` than `debug`? When do we use `debug` rather than `rel-debug`? It would be helpful to provide guidance here.

File: tutorials/holoscan_debugging/debug_legacy.sh [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+pushd ${tmp_dir}
+if [ ! -d holoscan-sdk ]; then
+    git clone git@github.com:nvidia-holoscan/holoscan-sdk.git
+fi
+cd holoscan-sdk
+git checkout ${holoscan_sdk_tag}
+# Apply patch to fix broken Vulkan SDK source
+# https://github.com/nvidia-holoscan/holoscan-sdk/issues/30
+git reset --hard HEAD
+git apply ${SCRIPT_DIR}/holoscan_sdk_20240723_1.diff
```
Comment (zhijinl) [inline tutorials/holoscan_debugging/debug_legacy.sh:40]: File "holoscan_sdk_20240723_1.diff" not found - forgot to include it in folder?

Comment (tbirdso) [reply tutorials/holoscan_debugging/debug_legacy.sh:40]: Apologies, missed this on check-in. Please try again now.

File: tutorials/holoscan_debugging/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+ARG BASE_IMAGE
+FROM ${BASE_IMAGE} as base
+
+RUN apt update && \
+    apt install -y gdb
```
Comment (zhijinl) [inline tutorials/holoscan_debugging/Dockerfile:22]: I tested `debug_legacy.sh` on my x86 machine, the `endoscopy_tool_tracking` app build process reports error - missing `ffmpeg`. 

After adding `ffmpeg` to `apt install -y`, all builds and runs without issue. Could you double check?

Comment (tbirdso) [reply tutorials/holoscan_debugging/Dockerfile:22]: Interesting, I'll look into this issue. @zhijinl could you please confirm whether you started from a clean build directory? If not, you can clear out your previous build with:
```
rm -rf build/endoscopy_tool_tracking
```

Comment (zhijinl) [reply tutorials/holoscan_debugging/Dockerfile:22]: Oh yes, you are correct - I wasn't building in a clean directory. I just tested a clean build, all works fine! Let's resolve this then

File: tutorials/holoscan_debugging/README.md (hunk 1 above)
Comment (tbirdso) [reply tutorials/holoscan_debugging/README.md:144]: Yes, will add a comment for clarity

File: tutorials/holoscan_debugging/README.md (hunk 5 above)
Comment (tbirdso) [reply tutorials/holoscan_debugging/README.md:249]: Due to https://github.com/nvidia-holoscan/holoscan-sdk/issues/30, the only legacy version that can be patched and built here is v2.2.0. Refactored the script and removed these arguments accordingly. (TBD whether we remove the legacy script as well.)

File: tutorials/holoscan_debugging/README.md [hunk 13]
```diff
@@ -0,0 +1,286 @@
...
+2. [Debugging a Python application with `pdb`](#debugging-a-python-application-with-pdb)
+3. [Debugging Symbols for Legacy Holoscan SDK](#debugging-symbols-for-legacy-holoscan-sdk-versions)
+4. [Logging](#logging)
+
+## Background
+
+### What is Debugging?
+
+Software debugging is the process of identifying and rectifying issues in software.
+- __Just-in-time debugging__ lets us pause execution and inspect state while a program is running.
```
[the remark this answers is not in the document]
Comment (tbirdso) [reply tutorials/holoscan_debugging/README.md]: Applied in latest batch

File: tutorials/holoscan_debugging/README.md (hunk 1 above)
Comment (tbirdso) [reply tutorials/holoscan_debugging/README.md]: Added 👍

File: tutorials/holoscan_debugging/README.md [hunk 14]
```diff
@@ -0,0 +1,302 @@
...
+## What are some common tools I can use for debugging my application?
+
+There are a wide variety of free and/or open source software tools available for general C++ and Python debugging, including:
+- [NVIDIA's debugging solutions](https://developer.nvidia.com/debugging-solutions), including NVIDIA NSight and CUDA-GDB;
+- The [GNU project DeBugger (GDB)]([GDB](https://sourceware.org/gdb/));
+- The built-in [Python Debugger (pdb)](https://docs.python.org/3/library/pdb.html) module;
+- [Microsoft Visual Studio Code](https://code.visualstudio.com/), with a wide variety of community extensions.
+
+In this tutorial we will focus on the "headless" `GDB` and `pdb` tools. These require minimal setup and can be run via a simple
+terminal without a dedicated display. For advanced development we recommend reviewing Visual Studio Code Development Containers
```
[the remark this answers is not in the document]
Comment (tbirdso) [reply tutorials/holoscan_debugging/README.md]: Unfortunately this suffers from the race condition where https://github.com/nvidia-holoscan/holohub/pull/430 adding VS Code support is not yet merged. I've updated this sentence to indicate that VS Code support is planned.

File: tutorials/holoscan_debugging/README.md (hunk 7 above)
Comment (tbirdso) [reply tutorials/holoscan_debugging/README.md]: Agreed, I've now moved these earlier in the tutorial as a general "Prerequisites" section

File: tutorials/holoscan_debugging/README.md (hunk 8 above)
Comment (tbirdso) [reply tutorials/holoscan_debugging/README.md]: Added

File: tutorials/holoscan_debugging/README.md (hunk 10 above)
Comment (tbirdso) [reply tutorials/holoscan_debugging/README.md]: In my view that discussion is better left in the introduction, added a sentence linking to that explanatory section.

File: tutorials/holoscan_debugging/Dockerfile (hunk 12 above)
Comment (tbirdso) [reply tutorials/holoscan_debugging/Dockerfile]: Added a sentence about clearing the build directory in the Prerequisites section, thanks for catching that!

File: tutorials/cli_debugging/Dockerfile [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+ARG BASE_IMAGE
+FROM ${BASE_IMAGE} as base
+
+RUN apt update && \
+    apt install -y gdb
```
Comment (maximilianofir) [inline tutorials/cli_debugging/Dockerfile:22]:
```suggestion
+   apt install -y gdb \
+    ffmpeg 
```
```

Comment (maximilianofir) [reply tutorials/cli_debugging/Dockerfile:22]: I needed to add ffmpeg on my x86 workstation, with HSDK 2.2

File: tutorials/cli_debugging/debug_legacy.sh [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+# Launch GDB with the Endoscopy Tool Tracking application
+${HOLOHUB_ROOT}/dev_container launch \
+    --docker_opts "-v ${INSTALL_DIR}:/opt/nvidia/holoscan --security-opt seccomp=unconfined" \
+    --img holohub:debugging \
+    --  bash -c \
+            'cd /workspace/holohub/build/endoscopy_tool_tracking && \
+            gdb -q \
+                -ex "break main" \
+                -ex "run --data /workspace/holohub/data/endoscopy" \
+                -ex "break /workspace/holoscan-sdk/src/core/application.cpp:add_flow" \
```
Comment (maximilianofir) [inline tutorials/cli_debugging/debug_legacy.sh:66]: when I launch the container seperately, holoscan-sdk is not mounted in /workspace. 
Oddly, the call here still works. 
Is there an explanation for this?

**Review by maximilianofir — COMMENTED** [review]
first review

File: tutorials/cli_debugging/Dockerfile (hunk 15 above)
Comment (tbirdso) [reply tutorials/cli_debugging/Dockerfile:22]: See @zhijinl 's comment -- would you please try again with a clear build cache? Though, I'm fine with adding `ffmpeg` to reduce recurring issues

Comment (tbirdso) [reply tutorials/cli_debugging/Dockerfile:22]: https://github.com/nvidia-holoscan/holohub/pull/433#discussion_r1691581916

File: tutorials/cli_debugging/debug_legacy.sh (hunk 16 above)
Comment (tbirdso) [reply tutorials/cli_debugging/debug_legacy.sh:66]: I'm having trouble reproducing. Here's what I'm seeing:

```
./tutorials/cli_debugging/debug_legacy.sh
# HSDK/container/app build, app launches, then stop container

./dev_container launch --img holohub:debugging

# In the container
cd /workspace/holohub/build/endoscopy_tool_tracking && \
            gdb -q \
                -ex "break main" \
                -ex "run --data /workspace/holohub/data/endoscopy" \
                -ex "break /workspace/holoscan-sdk/src/core/application.cpp:add_flow" \
/workspace/holohub/build/endoscopy_tool_tracking/applications/endoscopy_tool_tracking/cpp/endoscopy_tool_tracking
```

With that I see that the application was not able to load Holoscan SDK source file symbols such as for `/workspace/holoscan-sdk/src/core/application.cpp`:
```
Breakpoint 1, 0x00005a3b8fcdee40 in main ()
No source file named /workspace/holoscan-sdk/src/core/application.cpp.
Make breakpoint pending on future shared library load? (y or [n])
```

That behavior is expected, as the v2.2 base container does not include debug symbols with Holoscan SDK binaries in `/opt/nvidia/holoscan`.

Could you please confirm:
- What is the size of your Holoscan SDK binary folder? I would expect to see ~300MB for Release symbols, ~1.1GB for the custom RelWithDebInfo build:
```
$ du -sh /opt/nvidia/holoscan
331M    /opt/nvidia/holoscan
```
- Are you breaking on a `.h` or `.cpp` file? It is expected that you can break on `.h` files under the v2.2 distribution, those are included in the Holoscan SDK binary development package at `/opt/nvidia/holoscan` and symbols can be generated when you build the Endoscopy Tool Tracking app.

Comment (maximilianofir) [reply tutorials/cli_debugging/debug_legacy.sh:66]: for relaunching the container I used the following: 
`cd ~/repos/github/holohub/tutorials/cli_debugging
INSTALL_DIR=$(realpath $(find . -type d -name "install-*"))
cd ~/repos/github/holohub
./dev_container launch --docker_opts "-v $INSTALL_DIR:/opt/nvidia/holoscan --security-opt seccomp=unconfined" --img holohub:debugging `

inside the container I then do: 
`cd /workspace/holohub/build/endoscopy_tool_tracking`
`gdb -q \
    -ex "break main" \
    -ex "run --data /workspace/holohub/data/endoscopy" \
    -ex "break /workspace/holoscan-sdk/src/core/application.cpp:add_flow" \
    /workspace/holohub/build/endoscopy_tool_tracking/applications/endoscopy_tool_tracking/cpp/endoscopy_tool_tracking`
    
    So i do mount the holoscan install location, but its mapped to /opt/nvidia/holoscan. 
    It surprises me that the breakpoint is set for /workspace/holoscna-sdk. 
    `maxofir@6d4a536-lcedt:~/build/endoscopy_tool_tracking$ gdb -q     -ex "break main"     -ex "run --data /workspace/holohub/data/endoscopy"     -ex "break /workspace/holoscan-sdk/src/core/application.cpp:add_flow"     /workspace/holohub/build/endoscopy_tool_tracking/applications/endoscopy_tool_tracking/cpp/endoscopy_tool_tracking
Reading symbols from /workspace/holohub/build/endoscopy_tool_tracking/applications/endoscopy_tool_tracking/cpp/endoscopy_tool_tracking...
(No debugging symbols found in /workspace/holohub/build/endoscopy_tool_tracking/applications/endoscopy_tool_tracking/cpp/endoscopy_tool_tracking)
Breakpoint 1 at 0x13e40
Starting program: /workspace/holohub/build/endoscopy_tool_tracking/applications/endoscopy_tool_tracking/cpp/endoscopy_tool_tracking --data /workspace/holohub/data/endoscopy
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, 0x0000555555567e40 in main ()
Breakpoint 2 at 0x7ffff6ce6ef0: file /workspace/holoscan-sdk/src/core/application.cpp, line 170.
`

File: tutorials/cli_debugging/README.md [hunk 17]
```diff
@@ -0,0 +1,311 @@
...
+- `rel-debug`: Instructs the C++ compiler to optimize where possible _and_ generate debugging symbols. We use this build type to create developer-friendly binaries packaged in the Holoscan SDK NGC containers. Debug symbols may occasionally have inaccuracies due to release optimizations.
+- `debug`: Instructs the C++ compiler to minimize optimizations and prioritize debugging symbol accuracy.
+As of v2.3, we do not release Holoscan SDK builds of this type, but we do provide a `run` script to help
+developers generate their own `debug` builds on demand.
+
+## What are some common tools I can use for debugging my application?
+
+There are a wide variety of free and/or open source software tools available for general C++ and Python debugging, including:
+- [NVIDIA's debugging solutions](https://developer.nvidia.com/debugging-solutions), including NVIDIA NSight and CUDA-GDB;
+- The [GNU project DeBugger (GDB)]([GDB](https://sourceware.org/gdb/));
```
Comment (maximilianofir) [inline tutorials/cli_debugging/README.md:48]:
```suggestion
- The [GNU project DeBugger (GDB)](https://sourceware.org/gdb/);
```

File: tutorials/cli_debugging/debug_legacy.sh (hunk 16 above)
Comment (tbirdso) [reply tutorials/cli_debugging/debug_legacy.sh:66]: Ah, thanks for the extra details, I understand now. Debug symbols are embedded in Holoscan SDK libraries as absolute paths with respect to the build container. That is, interactive debugging will appear to occur with respect to the Holoscan SDK source path at build time, which was mounted at `/workspace/holoscan-sdk`.

That means there are three related but different directories to consider:
- Where the Holoscan SDK source is mounted at build time (`/workspace/holoscan-sdk`) determines debug symbol mappings
- Where the Holoscan SDK binaries are mounted at runtime (`/opt/nvidia/holoscan`) determines what binaries are available
- Where the Holoscan SDK source is mounted at runtime (`/workspace/holoscan-sdk`) has no direct impact on GDB debugging, but is useful for reference

TL;DR: This is expected behavior but can be easily misleading if the Holoscan SDK mount path changes in the run container. I'll add a note in this tutorial warning about that potential confusion.

File: tutorials/cli_debugging/Dockerfile (hunk 15 above)
Comment (tbirdso) [reply tutorials/cli_debugging/Dockerfile]: Added `ffmpeg` with an explanatory comment

File: tutorials/cli_debugging/debug_legacy.sh (hunk 16 above)
Comment (tbirdso) [reply tutorials/cli_debugging/debug_legacy.sh:66]: Add a small note in the README on file paths from debugging symbols.
